src/rust/src/x509.rs

```diff
@@ -43,6 +43,502 @@ lazy_static::lazy_static! {
     static ref CP_USER_NOTICE_OID: asn1::ObjectIdentifier<'static> = asn1::ObjectIdentifier::from_string("1.3.6.1.5.5.7.2.2").unwrap();
 }
 
+#[derive(asn1::Asn1Read, asn1::Asn1Write)]
+struct RawCertificate<'a> {
+    tbs_cert: TbsCertificate<'a>,
+    signature_alg: AlgorithmIdentifier<'a>,
+    signature: asn1::BitString<'a>,
+}
+
+#[derive(asn1::Asn1Read, asn1::Asn1Write)]
+struct TbsCertificate<'a> {
+    #[explicit(0)]
+    #[default(0)]
+    version: u8,
+    serial: asn1::BigUint<'a>,
+    _signature_alg: asn1::Sequence<'a>,
+
+    issuer: Name<'a>,
+    validity: Validity,
+    subject: Name<'a>,
+
+    spki: asn1::Sequence<'a>,
+    #[implicit(1)]
+    _issuer_unique_id: Option<asn1::BitString<'a>>,
+    #[implicit(2)]
+    _subject_unique_id: Option<asn1::BitString<'a>>,
+    #[explicit(3)]
+    extensions: Option<Extensions<'a>>,
+}
+
+pub(crate) type Name<'a> = asn1::SequenceOf<'a, asn1::SetOf<'a, AttributeTypeValue<'a>>>;
+
+#[derive(asn1::Asn1Read, asn1::Asn1Write)]
+pub(crate) struct AttributeTypeValue<'a> {
+    pub(crate) type_id: asn1::ObjectIdentifier<'a>,
+    pub(crate) value: asn1::Tlv<'a>,
+}
+
+#[derive(asn1::Asn1Read, asn1::Asn1Write)]
+enum Time {
+    UtcTime(asn1::UtcTime),
+    GeneralizedTime(asn1::GeneralizedTime),
+}
+
+impl Time {
+    fn as_chrono(&self) -> &chrono::DateTime<chrono::Utc> {
+        match self {
+            Time::UtcTime(data) => data.as_chrono(),
+            Time::GeneralizedTime(data) => data.as_chrono(),
+        }
+    }
+}
+
+#[derive(asn1::Asn1Read, asn1::Asn1Write)]
+pub(crate) struct Validity {
+    not_before: Time,
+    not_after: Time,
+}
+
+#[ouroboros::self_referencing]
+struct OwnedRawCertificate {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    value: RawCertificate<'this>,
+}
+
+#[pyo3::prelude::pyclass]
+struct Certificate {
+    raw: OwnedRawCertificate,
+    cached_extensions: Option<pyo3::PyObject>,
+}
+
+#[pyo3::prelude::pyproto]
+impl pyo3::class::basic::PyObjectProtocol for Certificate {
+    fn __hash__(&self) -> u64 {
+        let mut hasher = DefaultHasher::new();
+        self.raw.borrow_data().hash(&mut hasher);
+        hasher.finish()
+    }
+
+    fn __richcmp__(
+        &self,
+        other: pyo3::pycell::PyRef<Certificate>,
+        op: pyo3::class::basic::CompareOp,
+    ) -> pyo3::PyResult<bool> {
+        match op {
+            pyo3::class::basic::CompareOp::Eq => {
+                Ok(self.raw.borrow_data() == other.raw.borrow_data())
```

Comment:
Mmm, is comparing by bytes always correct, or are there ever cases where different bytes are really the same?

Reply:
https://github.com/openssl/openssl/blob/10af976962b2383bb3044120a764037361b8bff7/crypto/x509/x509_cmp.c#L155

OpenSSL generates a hash and verifies that the "encoding" hasn't changed. Digging through the layers it looks like byte for byte comparison probably is fine? It's also the case that the only bytes that could ever be different would be the ones outside of tbsCertificateBytes of course.